InterfaceFortran/TasmanianSG.f90

```diff
@@ -154,501 +166,646 @@ Module TasmanianSG
+end subroutine tsgSetAnisotropicRefinement
+!=======================================================================
+function tsgEstimateAnisotropicCoefficients(gridID, gtype, output) result(coeff)
+  integer :: gridID, gtype, output, N
+  integer, pointer :: coeff(:)
   N = tsgGetNumDimensions(gridID)
-  IF ((gtype .EQ. 2) .OR. (gtype .EQ. 4) .OR. (gtype .EQ. 6))then
+  if ((gtype .EQ. 2) .OR. (gtype .EQ. 4) .OR. (gtype .EQ. 6))then
     N = N * 2
-  ENDIF
-  ALLOCATE(coeff(N))
-  CALL tsgeac(gridID, gtype, output, coeff)
-END FUNCTION tsgEstimateAnisotropicCoefficients
-!=======================================================================
-SUBROUTINE tsgSetGlobalSurplusRefinement(gridID, tolerance, output)
-  INTEGER :: gridID, output
-  DOUBLE PRECISION :: tolerance
-  CALL tsgssr(gridID, tolerance, output)
-END SUBROUTINE tsgSetGlobalSurplusRefinement
-!=======================================================================
-SUBROUTINE tsgSetLocalSurplusRefinement(gridID, tolerance, rtype, output)
-  INTEGER :: gridID, rtype, theout
-  INTEGER, optional :: output
-  DOUBLE PRECISION :: tolerance
-  IF(PRESENT(output))then
-    theout = output
-  ELSE
+  endif
+  allocate(coeff(N))
+  call tsgeac(gridID, gtype, output-1, coeff)
+end function tsgEstimateAnisotropicCoefficients
+!=======================================================================
+subroutine tsgSetGlobalSurplusRefinement(gridID, tolerance, output, levelLimits)
+  integer :: gridID, output
+  integer, optional :: levelLimits(:)
+  double precision :: tolerance
+  if(present(levelLimits))then
+    call tsgssr(gridID, tolerance, output-1, levelLimits)
+  else
+    call tsgssr(gridID, tolerance, output-1)
+  endif
+end subroutine tsgSetGlobalSurplusRefinement
+!=======================================================================
+subroutine tsgSetLocalSurplusRefinement(gridID, tolerance, rtype, output, levelLimits)
+  integer :: gridID, rtype, theout
+  integer, optional :: output, levelLimits(:)
+  double precision :: tolerance
+  if(present(output))then
+    theout = output-1
+  else
     theout = -1
-  ENDIF
-  CALL tsgshr(gridID, tolerance, rtype, theout)
-END SUBROUTINE tsgSetLocalSurplusRefinement
-!=======================================================================
-SUBROUTINE tsgClearRefinement(gridID)
-  INTEGER :: gridID
-  CALL tsgcre(gridID)
-END SUBROUTINE tsgClearRefinement
-!=======================================================================
-SUBROUTINE tsgSetConformalTransformASIN(gridID, truncate)
-  INTEGER :: gridID, truncate(:)
-  CALL tsgsca(gridID, truncate)
-END SUBROUTINE tsgSetConformalTransformASIN
-!=======================================================================
-FUNCTION tsgIsSetConformalTransformASIN(gridID) result(isset)
-  INTEGER :: gridID, res
-  LOGICAL :: isset
-  CALL tsgica(gridID, res)
-  IF (res .EQ. 0)then
+  endif
+  if(present(levelLimits))then
+    call tsgshr(gridID, tolerance, rtype, theout, levelLimits)
```

Comment:
* take all pointers that have optional values, e.g., can be `null` and add a flag in the C wrappers
* if the flag is true, use the value provided by the Fortran module, if false pass to C++ the `null` pointer
* this will cover *make* functions and *refine* functions, and refine
* double-check to see if there are other functions that take default values